nistats/first_level_model.py

```diff
@@ -0,0 +1,606 @@
+            minimize_memory=False):
+    """ GLM fit for an fMRI data matrix
+
+    Parameters
+    ----------
+    Y : array of shape (n_time_points, n_voxels)
+        The fMRI data.
+
+    X : array of shape (n_time_points, n_regressors)
+        The design matrix.
+
+    noise_model : {'ar1', 'ols'}, optional
+        The temporal variance model. Defaults to 'ar1'.
+
+    bins : int, optional
+        Maximum number of discrete bins for the AR(1) coef histogram.
+
+    n_jobs : int, optional
+        The number of CPUs to use to do the computation. -1 means
+        'all CPUs'.
+
+    verbose : int, optional
+        The verbosity level. Defaut is 0
+
+    Returns
+    -------
+    labels : array of shape (n_voxels,),
+        A map of values on voxels used to identify the corresponding model.
+
+    results : dict,
+        Keys correspond to the different labels values
+        values are RegressionResults instances corresponding to the voxels.
+    """
+    acceptable_noise_models = ['ar1', 'ols']
+    if noise_model not in acceptable_noise_models:
+        raise ValueError(
+            "Acceptable noise models are {0}. You provided 'noise_model={1}'".\
+                format(acceptable_noise_models, noise_model))
+
+    if Y.shape[0] != X.shape[0]:
+        raise ValueError(
+            'The number of rows of Y should match the number of rows of X.'
+            ' You provided X with shape {0} and Y with shape {1}'.\
+                format(X.shape, Y.shape))
+
+    # Create the model
+    # model = OLSModel(X)
+    # Parallelize model fit by voxel groups
+    n_voxels = Y.shape[1]
+    group_iter = GroupIterator(n_voxels, n_jobs)
+    ols_result = Parallel(n_jobs=n_jobs, verbose=verbose)(
+        delayed(_OLSModel_fit)(X, Y[:, voxels]) for voxels in group_iter)
+    ols_result = concatenate_regression_results(ols_result)
+
+    if noise_model == 'ar1':
+        # compute and discretize the AR1 coefs
+        ar1 = ((ols_result.resid[1:] * ols_result.resid[:-1]).sum(axis=0) /
+               (ols_result.resid ** 2).sum(axis=0))
+        del ols_result
+        ar1 = (ar1 * bins).astype(np.int) * 1. / bins
+        # Fit the AR model acccording to current AR(1) estimates
+        results = {}
+        labels = ar1
+        # Parallelize by creating a job per ARModel
+        vals = np.unique(ar1)
+        ar_result = Parallel(n_jobs=n_jobs, verbose=verbose)(
+            delayed(_ARModel_fit)(X, val, Y[:, labels == val]) for val in vals)
+        for val, result in zip(vals, ar_result):
+            # We save memory if inspecting model details is not necessary
+            if minimize_memory:
+                result = _minimize_memory_regression_results(result)
+            results[val] = result
+
+    else:
+        labels = np.zeros(Y.shape[1])
+        results = {0.0: ols_result}
+
+    return labels, results
+
+
+def _check_list_length_match(list_1, list_2, var_name_1, var_name_2):
+    if len(list_1) != len(list_2):
```

Comment:
This function should have a (1-line) docstring and it should not be here, but in utils.py